src/compiler/build/build-stats.ts

```diff
@@ -0,0 +1,177 @@
+import { byteSize, sortBy } from '@utils';
+import type * as d from '../../declarations';
+
+/**
+ *
+ * @param config
+ * @param compilerCtx
+ * @param buildCtx
+ * @returns
+ */
+export function generateBuildStats(
+  config: d.Config,
+  buildCtx: d.BuildCtx
+): d.CompilerBuildStats | { diagnostics: d.Diagnostic[] } {
+  const buildResults = buildCtx.buildResults;
+
+  let jsonData: d.CompilerBuildStats | { diagnostics: d.Diagnostic[] };
+
+  try {
+    if (buildResults.hasError) {
+      jsonData = {
+        diagnostics: buildResults.diagnostics,
+      };
+    } else {
+      const stats: d.CompilerBuildStats = {
+        timestamp: buildResults.timestamp,
+        compiler: {
+          name: config.sys.name,
+          version: config.sys.version,
+        },
+        app: {
+          namespace: config.namespace,
+          fsNamespace: config.fsNamespace,
+          components: Object.keys(buildResults.componentGraph).length,
+          entries: Object.keys(buildResults.componentGraph).length,
+          bundles: buildResults.outputs.reduce((total, en) => (total += en.files.length), 0),
+          outputs: getAppOutputs(config, buildResults),
+        },
+        options: {
+          minifyJs: config.minifyJs,
+          minifyCss: config.minifyCss,
+          hashFileNames: config.hashFileNames,
+          hashedFileNameLength: config.hashedFileNameLength,
+          buildEs5: config.buildEs5,
+        },
+        formats: {
+          esmBrowser: sanitizeBundlesForStats(buildCtx.esmBrowserComponentBundle),
+          esm: sanitizeBundlesForStats(buildCtx.esmComponentBundle),
+          es5: sanitizeBundlesForStats(buildCtx.es5ComponentBundle),
+          system: sanitizeBundlesForStats(buildCtx.systemComponentBundle),
+          commonjs: sanitizeBundlesForStats(buildCtx.commonJsComponentBundle),
+        },
+        components: getComponentsFileMap(config, buildCtx),
+        entries: buildCtx.entryModules,
+        componentGraph: buildResults.componentGraph,
+        sourceGraph: getSourceGraph(config, buildCtx),
+        rollupResults: buildCtx.rollupResults,
+        collections: getCollections(config, buildCtx),
+      };
+
+      jsonData = stats;
+    }
+  } catch (e) {
+    jsonData = {
+      diagnostics: [e.message],
+    };
+  }
+
+  return jsonData;
+}
+
+/**
+ *
+ * @param config
+ * @param buildCtx
+ * @returns
+ */
+export async function writeBuildStats(config: d.Config, data: d.CompilerBuildStats | { diagnostics: d.Diagnostic[] }) {
+  const statsTargets = config.outputTargets.filter((o) => o.type === 'stats') as d.OutputTargetStats[];
+
+  await Promise.all(
+    statsTargets.map((outputTarget) => {
+      config.sys.writeFile(outputTarget.file, JSON.stringify(data, null, 2));
+    })
+  );
+}
+
+function sanitizeBundlesForStats(bundleArray: d.BundleModule[]): d.CompilerBuildStatBundle[] {
```

Comment:
Consider making this return `ReadonlyArray<d.BundleModule>`
```suggestion
function sanitizeBundlesForStats(bundleArray: ReadonlyArray<d.BundleModule>): ReadonlyArray<d.CompilerBuildStatBundle> {
```

although this will require some changes to how `CompilerBuildStats` is defined, which I've suggested in the file as well